core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/internals/network/NetworkErrorStatus.kt

```diff
@@ -1,70 +1,88 @@
+     * Indicates HTTP status code 413.
      */
-    ERROR_413,
+    data object Error413 : NetworkErrorStatus(HTTP_413)
 
     /**
-     * Indicates a retry able error, typically associated with HTTP status code 4xx-5xx, excluding other error listed above.
+     * Indicates a retry able error.
+     * This accepts a dynamic error code for cases when the code is not one of the predefined ones.
+     *
+     * The value will be null if the error code is not available e.g., in case of IO exception.
+     *
+     * @param responseCode The HTTP status code associated with the error, if applicable.
      */
-    ERROR_RETRY,
+    data class ErrorRetry(override val responseCode: Int? = null) : NetworkErrorStatus(responseCode)
 
     /**
-     * Indicates a retry able error, typically happens when the network is unavailable.
+     * Indicates a retry able error that happens when the network is unavailable.
      */
-    ERROR_NETWORK_UNAVAILABLE,
+    data object ErrorNetworkUnavailable : NetworkErrorStatus(null)
 
     /**
-     * Indicates a fatal error, typically associated with some exception or failure that cannot be retried.
+     * Indicates an unknown but retry able error.
      */
-    ERROR_UNKNOWN;
+    data object ErrorUnknown : NetworkErrorStatus(null)
 
     companion object {
 
         /**
-         * Converts an HTTP status code to a corresponding `ErrorStatus` enum value.
+         * Converts an HTTP status code to a corresponding NetworkErrorStatus instance.
          *
-         * This method maps HTTP status codes and other specific error codes to the appropriate `ErrorStatus` enum values
-         * for easier handling of different error conditions.
-         *
-         * @param errorCode The HTTP status code or error code to be mapped to an `ErrorStatus`.
-         * @return The corresponding `ErrorStatus` enum value.
+         * @param errorCode The HTTP status code to be mapped.
+         * @return The corresponding NetworkErrorStatus instance.
          */
         fun toErrorStatus(errorCode: Int): NetworkErrorStatus = when (errorCode) {
-            BAD_REQUEST_CODE -> ERROR_400
-            UNAUTHORIZED_CODE -> ERROR_401
-            RESOURCE_NOT_FOUND_CODE -> ERROR_404
-            PAYLOAD_TOO_LARGE_CODE -> ERROR_413
-            else -> ERROR_RETRY
+            HTTP_400 -> Error400
+            HTTP_401 -> Error401
+            HTTP_404 -> Error404
+            HTTP_413 -> Error413
+            in HTTP_RETRY_RANGE -> ErrorRetry(errorCode)
+            else -> ErrorUnknown
         }
     }
 }
+
+/**
+ * Extension function to format the response code message for a NetworkErrorStatus.
+ * @return A string representation of the response code, or "Not available" if the code is null.
+ */
+internal fun NetworkErrorStatus.formatResponseCodeMessage(): String {
+    val responseCode = this.responseCode ?: "Not available"
+    return "Response code: $responseCode"
```

Comment:
It's actually called "Status code" (short for HTTP Response Status Code). So, you may want to correct this convention everywhere.